### PR TITLE
Log part 4.1 (?) Searched text highlighting

### PR DESCRIPTION
**Pull request by ricdezen**
## Premise

As always, please see part 4 first.
This is not all I wanted to put in the next pull request, but I decided to give priority to other stuff.

## Changes

- Added proper SearchView implementation in the fragment
- LogItems can now be filtered on both name and address
- LogItems implement a new interface called `Markable` that identifies an item that can be marked based on some kind of criteria. For LogItem this criteria is a String, if such a String is found in the name or address, the text is highlighted in a different color.
- `FilterList` is now an interface, `LogList` is a class that implements both `FilterList` and `Markable`.
- Added some tests here and there. All my classes except `LogRecyclerAdapter` have now very close to 100% coverage, although due to Robolectric some workarounds need to be done to get the correct report.

As always you can see the result inside LogFragmentTest.

### Footer

Trying to stay positive. Have an... I dunno a [duck](https://sm.ign.com/t/ign_it/blogroll/h/heres-what/heres-what-untitled-goose-game-was-almost-called_v9qn.1280.jpg) whatever thanks for your time.

Jokes on you it's actually a goose.

### CODE REVIEW

File: app/src/main/java/ingsw/group1/findmyphone/fragment/LogFragment.java [hunk 1]
```diff
@@ -62,11 +68,108 @@ public View onCreateView(
...
+
+    /**
+     * Method called to inflate a popup menu and attach it to a View.
+     * Needs to be called by passing {@link LogFragment#sortButton} and for the fragment's
+     * {@link Context} to not be {@code null} in order to proceed.
+     *
+     * @param anchorView The view to which the menu will be attached.
+     */
+    public void showMenu(View anchorView) {
+        if (getContext() == null || anchorView != sortButton) return;
```
Comment (RicDeZen) [inline app/src/main/java/ingsw/group1/findmyphone/fragment/LogFragment.java]: Codacy Issue found: [Use equals() to compare object references.](https://app.codacy.com/manual/RicDeZen/ingsw-group1-findmyphone/pullRequest?prid=5028692)

**Comment by RicDeZen** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Complexity increasing per file
==============================
- app/src/main/java/ingsw/group1/findmyphone/log/LogItemComparatorHelper.java  4
- app/src/main/java/ingsw/group1/findmyphone/log/LogItemFormatter.java  4
- app/src/main/java/ingsw/group1/findmyphone/log/LogManager.java  5
- app/src/testShared/java/ingsw/group1/findmyphone/random/RandomGeoPositionGenerator.java  1
- app/src/testShared/java/ingsw/group1/findmyphone/random/RandomSMSLogEventGeneratorTest.java  1
- app/src/testShared/java/ingsw/group1/findmyphone/random/RandomLogItemGenerator.java  2
- app/src/test/java/ingsw/group1/findmyphone/log/LogItemSpansTest.java  1
- app/src/test/java/ingsw/group1/findmyphone/log/LogItemFormatterTest.java  2
- app/src/test/java/ingsw/group1/findmyphone/log/LogManagerTest.java  4
- app/src/main/java/ingsw/group1/findmyphone/log/LogItem.java  2
- app/src/main/java/ingsw/group1/findmyphone/log/LogList.java  3
- app/src/test/java/ingsw/group1/findmyphone/log/LogItemComparatorHelperTest.java  1
- app/src/main/java/ingsw/group1/findmyphone/fragment/LogFragment.java  4
- app/src/testShared/java/ingsw/group1/findmyphone/random/RandomSMSLogEventGenerator.java  5
- app/src/testShared/java/ingsw/group1/findmyphone/random/RandomSMSContactGenerator.java  1
- app/src/test/java/ingsw/group1/findmyphone/log/LogViewHolderTest.java  3
- app/src/test/java/ingsw/group1/findmyphone/log/LogListTest.java  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/RicDeZen/ingsw-group1-findmyphone/pullRequest?prid=5028692&bid=16570766)

File: app/src/main/java/ingsw/group1/findmyphone/fragment/LogFragment.java [hunk 2]
```diff
@@ -3,31 +3,37 @@
...
  * @author Riccardo De Zen.
  */
-public class LogFragment extends Fragment {
+public class LogFragment extends Fragment implements PopupMenu.OnMenuItemClickListener {
 
-    private List<LogItem> logItemList;
+    private LogManager logManager;
+    private RecyclerView logRecycler;
+    private ImageButton sortButton;
+    private SearchView searchView;
```
Comment (RicDeZen) [inline app/src/main/java/ingsw/group1/findmyphone/fragment/LogFragment.java]: Codacy Issue found: [Perhaps 'searchView' could be replaced by a local variable.](https://app.codacy.com/manual/RicDeZen/ingsw-group1-findmyphone/pullRequest?prid=5028692)

File: app/src/main/java/ingsw/group1/findmyphone/fragment/LogFragment.java [hunk 3]
```diff
@@ -62,11 +68,108 @@ public View onCreateView(
...
+
+    /**
+     * Method called when a menu item is clicked.
+     *
+     * @param item The clicked item.
+     * @return {@code true} if the action was managed, {@code false} otherwise.
+     */
+    @Override
+    public boolean onMenuItemClick(MenuItem item) {
+        AdapterView.AdapterContextMenuInfo info =
```
Comment (RicDeZen) [inline app/src/main/java/ingsw/group1/findmyphone/fragment/LogFragment.java]: Codacy Issue found: [Avoid unused local variables such as 'info'.](https://app.codacy.com/manual/RicDeZen/ingsw-group1-findmyphone/pullRequest?prid=5028692)

**Comment by RicDeZen** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
- Added 3
           

Complexity increasing per file
==============================
- app/src/main/java/ingsw/group1/findmyphone/log/LogManager.java  5
- app/src/test/java/ingsw/group1/findmyphone/log/LogManagerTest.java  4
- app/src/main/java/ingsw/group1/findmyphone/log/LogItemComparatorHelper.java  4
- app/src/testShared/java/ingsw/group1/findmyphone/random/RandomSMSLogEventGeneratorTest.java  1
- app/src/test/java/ingsw/group1/findmyphone/log/LogListTest.java  4
- app/src/testShared/java/ingsw/group1/findmyphone/random/RandomGeoPositionGenerator.java  1
- app/src/test/java/ingsw/group1/findmyphone/log/LogViewHolderTest.java  3
- app/src/testShared/java/ingsw/group1/findmyphone/random/RandomLogItemGenerator.java  2
- app/src/test/java/ingsw/group1/findmyphone/log/LogItemFormatterTest.java  2
- app/src/test/java/ingsw/group1/findmyphone/log/LogItemComparatorHelperTest.java  1
- app/src/main/java/ingsw/group1/findmyphone/log/LogList.java  3
- app/src/main/java/ingsw/group1/findmyphone/log/LogItem.java  2
- app/src/testShared/java/ingsw/group1/findmyphone/random/RandomSMSLogEventGenerator.java  5
- app/src/main/java/ingsw/group1/findmyphone/log/LogItemFormatter.java  4
- app/src/testShared/java/ingsw/group1/findmyphone/random/RandomSMSContactGenerator.java  1
- app/src/test/java/ingsw/group1/findmyphone/log/LogItemSpansTest.java  1
- app/src/main/java/ingsw/group1/findmyphone/fragment/LogFragment.java  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/RicDeZen/ingsw-group1-findmyphone/pullRequest?prid=5028692&bid=16570766)

File: app/src/main/java/ingsw/group1/findmyphone/fragment/LogFragment.java (hunk 2 above)
Comment (RicDeZen) [inline app/src/main/java/ingsw/group1/findmyphone/fragment/LogFragment.java]: Codacy Issue found: [Perhaps 'searchView' could be replaced by a local variable.](https://app.codacy.com/manual/RicDeZen/ingsw-group1-findmyphone/pullRequest?prid=5028692)

File: app/src/main/java/ingsw/group1/findmyphone/fragment/LogFragment.java (hunk 1 above)
Comment (RicDeZen) [inline app/src/main/java/ingsw/group1/findmyphone/fragment/LogFragment.java]: Codacy Issue found: [Use equals() to compare object references.](https://app.codacy.com/manual/RicDeZen/ingsw-group1-findmyphone/pullRequest?prid=5028692)

File: app/src/main/java/ingsw/group1/findmyphone/fragment/LogFragment.java (hunk 3 above)
Comment (RicDeZen) [inline app/src/main/java/ingsw/group1/findmyphone/fragment/LogFragment.java]: Codacy Issue found: [Avoid unused local variables such as 'info'.](https://app.codacy.com/manual/RicDeZen/ingsw-group1-findmyphone/pullRequest?prid=5028692)

**Comment by RicDeZen** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Complexity increasing per file
==============================
- app/src/main/java/ingsw/group1/findmyphone/log/LogItem.java  2
- app/src/test/java/ingsw/group1/findmyphone/log/LogManagerTest.java  4
- app/src/test/java/ingsw/group1/findmyphone/log/LogItemSpansTest.java  1
- app/src/test/java/ingsw/group1/findmyphone/log/LogItemFormatterTest.java  2
- app/src/androidTest/java/ingsw/group1/findmyphone/AndroidTestUtils.java  1
- app/src/test/java/ingsw/group1/findmyphone/log/LogListTest.java  4
- app/src/main/java/ingsw/group1/findmyphone/fragment/SettingsFragment.java  1
- app/src/main/java/ingsw/group1/findmyphone/log/LogList.java  3
- app/src/main/java/ingsw/group1/findmyphone/fragment/LogFragment.java  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/RicDeZen/ingsw-group1-findmyphone/pullRequest?prid=5028692&bid=16570766)

File: app/src/main/java/ingsw/group1/findmyphone/fragment/LogFragment.java [hunk 4]
```diff
@@ -59,12 +69,109 @@ public View onCreateView(
...
+
+    /**
+     * Method called to inflate a popup menu and attach it to a View.
+     * Needs to be called by passing {@link LogFragment#sortButton} and for the fragment's
+     * {@link Context} to not be {@code null} in order to proceed.
+     *
+     * @param anchorView The view to which the menu will be attached.
+     */
+    public void showMenu(View anchorView) {
+        if (getContext() == null || anchorView != sortButton) return;
```
Comment (RicDeZen) [inline app/src/main/java/ingsw/group1/findmyphone/fragment/LogFragment.java]: Codacy Issue found: [Use equals() to compare object references.](https://app.codacy.com/manual/RicDeZen/ingsw-group1-findmyphone/pullRequest?prid=5028692)

File: app/src/main/java/ingsw/group1/findmyphone/fragment/LogFragment.java [hunk 5]
```diff
@@ -59,12 +69,109 @@ public View onCreateView(
...
+
+    /**
+     * Method called when a menu item is clicked.
+     *
+     * @param item The clicked item.
+     * @return {@code true} if the action was managed, {@code false} otherwise.
+     */
+    @Override
+    public boolean onMenuItemClick(MenuItem item) {
+        AdapterView.AdapterContextMenuInfo info =
```
Comment (RicDeZen) [inline app/src/main/java/ingsw/group1/findmyphone/fragment/LogFragment.java]: Codacy Issue found: [Avoid unused local variables such as 'info'.](https://app.codacy.com/manual/RicDeZen/ingsw-group1-findmyphone/pullRequest?prid=5028692)

File: app/src/main/java/ingsw/group1/findmyphone/fragment/LogFragment.java [hunk 6]
```diff
@@ -22,9 +28,12 @@
  *
  * @author Riccardo De Zen.
  */
-public class LogFragment extends Fragment {
+public class LogFragment extends Fragment implements PopupMenu.OnMenuItemClickListener {
 
     private LogManager logManager;
+    private RecyclerView logRecycler;
+    private ImageButton sortButton;
+    private SearchView searchView;
```
Comment (RicDeZen) [inline app/src/main/java/ingsw/group1/findmyphone/fragment/LogFragment.java]: Codacy Issue found: [Perhaps 'searchView' could be replaced by a local variable.](https://app.codacy.com/manual/RicDeZen/ingsw-group1-findmyphone/pullRequest?prid=5028692)

**Comment by RicDeZen** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- app/src/test/java/ingsw/group1/findmyphone/log/LogItemSpansTest.java  1
- app/src/androidTest/java/ingsw/group1/findmyphone/AndroidTestUtils.java  1
- app/src/main/java/ingsw/group1/findmyphone/log/LogList.java  3
- app/src/main/java/ingsw/group1/findmyphone/fragment/SettingsFragment.java  1
- app/src/main/java/ingsw/group1/findmyphone/log/LogItem.java  2
- app/src/test/java/ingsw/group1/findmyphone/log/LogListTest.java  4
- app/src/test/java/ingsw/group1/findmyphone/log/LogManagerTest.java  4
- app/src/test/java/ingsw/group1/findmyphone/log/LogItemFormatterTest.java  2
- app/src/main/java/ingsw/group1/findmyphone/fragment/LogFragment.java  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/RicDeZen/ingsw-group1-findmyphone/pullRequest?prid=5028692&bid=16570766)

File: app/src/main/java/ingsw/group1/findmyphone/fragment/LogFragment.java [hunk 7]
```diff
@@ -59,12 +68,116 @@ public View onCreateView(
...
+
         return root;
     }
+
+    /**
+     * Method to submit a query to the SearchView programmatically.
+     *
+     * @param query The query to submit.
+     */
+    private void pushQuery(String query) {
```
Comment (RicDeZen) [inline app/src/main/java/ingsw/group1/findmyphone/fragment/LogFragment.java]: Codacy Issue found: [Avoid unused private methods such as 'pushQuery(String)'.](https://app.codacy.com/manual/RicDeZen/ingsw-group1-findmyphone/pullRequest?prid=5028692)

Comment (RicDeZen) [reply app/src/main/java/ingsw/group1/findmyphone/fragment/LogFragment.java]: Ma puppami la fava Codacy l'ho messo apposta perchè scassavi che searchView andava messa variabile locale non sei mai contento

**Comment by RicDeZen** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- app/src/test/java/ingsw/group1/findmyphone/log/LogItemSpansTest.java  1
- app/src/androidTest/java/ingsw/group1/findmyphone/AndroidTestUtils.java  1
- app/src/test/java/ingsw/group1/findmyphone/log/LogManagerTest.java  4
- app/src/test/java/ingsw/group1/findmyphone/log/LogListTest.java  4
- app/src/main/java/ingsw/group1/findmyphone/log/LogList.java  3
- app/src/test/java/ingsw/group1/findmyphone/log/LogItemFormatterTest.java  2
- app/src/main/java/ingsw/group1/findmyphone/fragment/SettingsFragment.java  1
- app/src/main/java/ingsw/group1/findmyphone/log/LogItem.java  2
- app/src/main/java/ingsw/group1/findmyphone/fragment/LogFragment.java  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/RicDeZen/ingsw-group1-findmyphone/pullRequest?prid=5028692&bid=16570766)

**Comment by ricdezen** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- app/src/main/java/ingsw/group1/findmyphone/log/LogList.java  3
- app/src/main/java/ingsw/group1/findmyphone/log/LogItem.java  2
- app/src/test/java/ingsw/group1/findmyphone/log/LogItemFormatterTest.java  2
- app/src/test/java/ingsw/group1/findmyphone/log/LogListTest.java  4
- app/src/test/java/ingsw/group1/findmyphone/log/LogItemSpansTest.java  1
- app/src/main/java/ingsw/group1/findmyphone/fragment/LogFragment.java  4
- app/src/test/java/ingsw/group1/findmyphone/log/LogManagerTest.java  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/RicDeZen/ingsw-group1-findmyphone/pullRequest?prid=5028692&bid=16570766)

**Review by turcato-niccolo — APPROVED** [review]

**Pull request merged**